### PR TITLE
Support sending and receiving multiple Justifications

### LINKED ISSUE #8172: Support sending multiple Justifications
The PR #7640 implements the first part of #7556 by adding support for storing multiple Justifications per block, but still maintains backwards compatibility by only sending single Justifications across the network, and only from GRANDPA. This issue tracks the follow up work that needs to be done, which is to also support sending/receiving multiple Justifications on the wire.

The approach is the one outlined in #7556 where we introduce another field `justifications`, as well as a field indicating if the requester supports multiple Justifications.

### PR DESCRIPTION
**Pull request by octol**
Follow up to #7640 , where we add support for also sending multiple Justifications across the network. We include in the request if we support receiving multiple at once, and are backwards compatible with old nodes.

Resolves #8172

### CODE REVIEW

**Review by andresilva — COMMENTED** [review]
Overall the changes LGTM but I'd like to make sure this is adequately tested to avoid any breakage.

**Comment by andresilva** [comment]
What's the difference between #7556 and #8172? Isn't #8172 a subset of #7556?

**Comment by octol** [comment]
> What's the difference between #7556 and #8172? Isn't #8172 a subset of #7556?

#8172 was supposed to track the pieces specific to the networking and also the remaining work needed. It's the one that the code refers to in `client/network` when the adding the GRANDPA consensus engine ID tag

**Comment by octol** [comment]
> Overall the changes LGTM but I'd like to make sure this is adequately tested to avoid any breakage.

Making a push to properly test this so we can hopefully merge it soon

File: client/network/src/protocol/sync.rs
Comment (octol) [inline client/network/src/protocol/sync.rs:453]: Been testing this against older nodes and I'm getting an issue where the older nodes fail to decode `BlockAttributes` in a request with the extra bit set

Comment (octol) [reply client/network/src/protocol/sync.rs:453]: Switched this logic from `BlockAttributes` to the protobuf schema instead

File: client/network/src/protocol.rs
Comment (tomaka) [inline client/network/src/protocol.rs:584]:
```suggestion
					Some(DecodeAll::decode_all(&mut block_data.justifications.as_ref())?)
```

File: client/network/src/schema/api.v1.proto
Comment (tomaka) [inline client/network/src/schema/api.v1.proto:65]: I don't really understand this comment.

**Review by tomaka — CHANGES_REQUESTED** [review]
Looks good

File: client/network/src/schema/api.v1.proto
Comment (octol) [reply client/network/src/schema/api.v1.proto]: Had another go at reformulating it to try to make it a bit more readable :)

**Review by tomaka — APPROVED** [review]

**Review by andresilva — APPROVED** [review]

**Comment by andresilva** [comment]
@octol Just to be sure before I click the merge button: when you tested this locally you tested both scenarios right? I.e. new nodes serving old nodes, and old nodes serving new nodes.

**Comment by octol** [comment]
> @octol Just to be sure before I click the merge button: when you tested this locally you tested both scenarios right? I.e. new nodes serving old nodes, and old nodes serving new nodes.

I tested having two nodes built using this PR finalizing blocks, and then a third and forth node connecting with `--no-grandpa`, one based on `master` and one on this PR, confirming that they are both able to follow finality (at sync and when the set id changes).

And then I redid this with the same setup but the nodes finalizing blocks were built using `master`.

This should cover both scenarios I think

**Comment by andresilva** [comment]
@octol Thanks!

bot merge

**Comment by andresilva** [comment]
bot merge

**Pull request merged**
